Python/optimizer_analysis.c

```diff
@@ -9,10 +20,457 @@
+    return null_sym;
+}
+
+
+static inline bool
+sym_matches_pytype(_Py_UOpsSymType *sym, PyTypeObject *typ)
+{
+    assert(typ == NULL || PyType_Check(typ));
+    if (!sym_has_flag(sym, KNOWN_TYPE)) {
+        return false;
+    }
+    return sym->typ == typ;
+}
+
+
+static inline bool
+op_is_end(uint32_t opcode)
+{
+    return opcode == _EXIT_TRACE || opcode == _JUMP_TO_TOP;
+}
+
+
+static inline bool
+op_is_bookkeeping(uint32_t opcode) {
+    return (opcode == _SET_IP ||
+            opcode == _CHECK_VALIDITY ||
+            opcode == _SAVE_RETURN_OFFSET ||
+            opcode == _RESUME_CHECK);
+}
+
+static inline bool
+op_is_data_movement_only(uint32_t opcode) {
+    return (opcode == _STORE_FAST ||
+        opcode == STORE_FAST_MAYBE_NULL ||
+        opcode == _PUSH_FRAME ||
+        opcode == _POP_FRAME);
+}
+
+
+static inline int
+emit_i(uops_emitter *emitter,
+       _PyUOpInstruction inst)
+{
+    if (emitter->writebuffer + emitter->curr_i >= emitter->writebuffer_end) {
+        OPT_STAT_INC(optimizer_failure_reason_no_writebuffer);
+        DPRINTF(1, "out of emission space\n");
+        return -1;
+    }
+    if (inst.opcode == _NOP) {
+        return 0;
+    }
+    DPRINTF(2, "Emitting instruction at [%d] op: %s, oparg: %d, target: %d, operand: %" PRIu64 " \n",
+            emitter->curr_i,
+            _PyOpcode_uop_name[inst.opcode],
+            inst.oparg,
+            inst.target,
+            inst.operand);
+    emitter->writebuffer[emitter->curr_i] = inst;
+    emitter->curr_i++;
+    return 0;
+}
+
+static int
+uop_abstract_interpret_single_inst(
+    _PyUOpInstruction *inst,
+    _PyUOpInstruction *end,
+    _Py_UOpsAbstractInterpContext *ctx
+)
+{
+#define STACK_LEVEL()     ((int)(stack_pointer - ctx->frame->stack))
+
+#define GETLOCAL(idx)          ((ctx->frame->locals[idx]))
+
+#define REPLACE_OP(op, arg, oper)    \
+    new_inst.opcode = op;            \
+    new_inst.oparg = arg;            \
+    new_inst.operand = oper;
+
+#define _LOAD_ATTR_NOT_NULL \
+    do {                    \
+    attr = sym_new_known_notnull(ctx); \
+    if (attr == NULL) { \
+        goto error; \
+    } \
+    null = sym_new_null(ctx); \
+    if (null == NULL) { \
+        goto error; \
+    } \
+    } while (0);
+
+    int oparg = inst->oparg;
+    uint32_t opcode = inst->opcode;
+
+    _Py_UOpsSymType **stack_pointer = ctx->frame->stack_pointer;
+    _PyUOpInstruction new_inst = *inst;
+
+    DPRINTF(3, "Abstract interpreting %s:%d ",
+            _PyOpcode_uop_name[opcode],
+            oparg);
+    switch (opcode) {
+#include "tier2_redundancy_eliminator_cases.c.h"
+
+        default:
+            DPRINTF(1, "Unknown opcode in abstract interpreter\n");
+            Py_UNREACHABLE();
+    }
+    assert(ctx->frame != NULL);
+    DPRINTF(3, " stack_level %d\n", STACK_LEVEL());
+    ctx->frame->stack_pointer = stack_pointer;
+    assert(STACK_LEVEL() >= 0);
+
+    if (emit_i(&ctx->emitter, new_inst) < 0) {
```

Comment:
Why are we copying every instruction? Modifying the instructions in place would give much better cache locality and saves a lot of work when instructions do not need to be modified.

I think you can delete `emit_i` and the associated data structures.

Reply:
When we do constant propagation we need to adjust the stack, in certain cases, so the extra stack adjustment instruction might cause the current instruction pointer to overtake the current one.

However, that is IF we implement it in that fashion, and I don't think we will. So yeah lets just remove it.